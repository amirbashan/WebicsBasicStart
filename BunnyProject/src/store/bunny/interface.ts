export interface Bunny {
  id: number;
  name: string;
  color: string;
  cuteness: number;
}

export interface colorChoice {
  [key: number]: string;
}

export interface Store {
  bunnies: Bunny[];
  loading:boolean,
  error:string
}
