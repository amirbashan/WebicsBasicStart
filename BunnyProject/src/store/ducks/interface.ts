export interface Duck {
  id: number;
  name: string;
  color: string;
  cuteness: number;
}

export interface colorChoice {
  [key: number]: string;
}

export interface Store {
  ducks: Duck[];
  loading:boolean,
  error:string
}
