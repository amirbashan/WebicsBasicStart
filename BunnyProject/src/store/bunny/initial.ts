import { Store } from "./interface";

export const initialState: Store = {
    bunnies: [],
    loading:false,
    error:""
}
