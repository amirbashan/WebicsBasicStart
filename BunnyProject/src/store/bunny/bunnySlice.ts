import { createSlice, PayloadAction, createAsyncThunk } from "@reduxjs/toolkit";
import { URL } from "../../App";
import { Bunny } from "./interface";
import { initialState } from "./initial";
import { randomNumber, colorIndex } from "../../util";
import axios from "axios";

export const getAllBunnies = createAsyncThunk("bunnies/getAllBunnies", async () => {
  const response = await axios.get(`${URL}/bunnies`);
  return response.data.rows;
});
export const addBunnyToDB = createAsyncThunk("bunnies/getAllBunnies", async (bunny:Bunny) => {
  const response = await axios.post(`${URL}/bunnies`,bunny);
  console.log(response)
  return response.data;
});

const bunnySlice = createSlice({
  name: "bunnies",
  initialState,
  reducers: {
    addBunny(state) {
      const length: number = state.bunnies.length;
      const nextId: number = length ? state.bunnies[length - 1].id + 1 : 1;
      const newBunny: Bunny = {
        id: nextId,
        name: "new Rabbit",
        cuteness: randomNumber(),
        color: colorIndex[randomNumber()],
      };
      addBunnyToDB(newBunny)
      state.bunnies.push(newBunny);
    },
    deleteBunny(state, action: PayloadAction<string>) {
      const newArray = state.bunnies.filter(
        (bunny) => bunny.id !== parseInt(action.payload)
      );
      return { ...state, bunnies: newArray };
    },
  },
  extraReducers: (builder) => {
    builder.addCase(getAllBunnies.pending, (state, action) => {
      state.loading = true;
    });
    builder.addCase(getAllBunnies.fulfilled, (state, action) => {
      state.loading = false;
      state.bunnies = action.payload;
      state.error = "";
    });
    builder.addCase(getAllBunnies.rejected, (state, action) => {
      state.loading = false;
      state.bunnies = [];
      state.error = action.error.message as string;
    });
  },
});

// export const bunniesArray = (state: any)=> state.bunnies

export const { addBunny, deleteBunny } = bunnySlice.actions;
export default bunnySlice.reducer;
