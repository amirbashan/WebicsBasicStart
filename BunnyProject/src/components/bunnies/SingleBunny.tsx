import React from "react";
import { Bunny } from "../../store/bunny/interface";

interface AppProps {
  bunny: Bunny;
}

export default function SingleBunny({ bunny }: AppProps) {
  return (
    <li>
      <div>Id: {bunny.id}</div>
      <div>Name: {bunny.name}</div>
      <div>Cuteness: {bunny.cuteness}</div>
      <div>Color: {bunny.color}</div>
    </li>
  );
}
