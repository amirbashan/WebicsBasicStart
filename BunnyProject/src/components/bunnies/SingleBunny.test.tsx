import React from "react";
import { expect, describe, test } from "@jest/globals";
import SingleBunny from "./SingleBunny";
import renderer from "react-test-renderer";
import { Bunny } from "../../store/bunny/interface";

describe("singleBunny", () => {
  test("Snapshot test", () => {
    const tree = renderer
      .create(
        <SingleBunny
          bunny={{ id: 1, name: "max", cuteness: 1, color: "white" } as Bunny}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
