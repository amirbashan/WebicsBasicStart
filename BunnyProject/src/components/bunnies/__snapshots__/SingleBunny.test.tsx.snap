// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`singleBunny Snapshot test 1`] = `
<li>
  <div>
    Id: 
    1
  </div>
  <div>
    Name: 
    max
  </div>
  <div>
    Cuteness: 
    1
  </div>
  <div>
    Color: 
    white
  </div>
</li>
`;
